Consuming Abseil via Protobuf as transitive dependency is fragile

Hello!

We have packaged libprotobuf-mutator in Conan (see PR [#24163](https://github.com/conan-io/conan-center-index/pull/24163)). During this process, we noticed that the [Abseil](https://github.com/abseil/abseil-cpp) library is not only used by Protobuf but also directly by libprotobuf-mutator. For instance, the [absl::StrCat](https://github.com/google/libprotobuf-mutator/blob/1f95f8083066f5b38fd2db172e7e7f9aa7c49d2d/src/field_instance.h#L198) function is used without a direct reference, such as a header include or CMake target.

This creates a fragile situation where the functionality may depend on Protobuf’s use of Abseil, leading to potential issues during linking. It would be beneficial to refactor the CMake configuration of this project to explicitly consume both Protobuf and Abseil. This can be achieved not only by using CMake targets but also by utilizing the official [FindProtobuf](https://cmake.org/cmake/help/latest/module/FindProtobuf.html) module to avoid additional overhead.

I am willing to submit a PR with these changes, but I would first like to hear your thoughts on this matter.

For proof that Abseil 20230125 is not recent enough since at least libprotobuf-mutator 1.3 with pull request #243 included:

```console
# pkg-config protobuf --modversion
22.5.0

# pkg-config absl_strings --modversion
20230125

# /usr/bin/c++ -dumpversion
13

# git rev-parse HEAD
5595eff26a93c91d13e8c073bd0de2b9f96d7693

# make VERBOSE=1
/usr/bin/cmake -S/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator -B/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/CMakeFiles /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator//CMakeFiles/progress.marks
make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator'
make  -f src/CMakeFiles/protobuf-mutator.dir/build.make src/CMakeFiles/protobuf-mutator.dir/depend
make[2]: Entering directory '/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator'
cd /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/CMakeFiles/protobuf-mutator.dir/DependInfo.cmake "--color="
Dependencies file "src/CMakeFiles/protobuf-mutator.dir/binary_format.cc.o.d" is newer than depends file "/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/CMakeFiles/protobuf-mutator.dir/compiler_depend.internal".
Dependencies file "src/CMakeFiles/protobuf-mutator.dir/mutator.cc.o.d" is newer than depends file "/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/CMakeFiles/protobuf-mutator.dir/compiler_depend.internal".
Consolidate compiler generated dependencies of target protobuf-mutator
make[2]: Leaving directory '/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator'
make  -f src/CMakeFiles/protobuf-mutator.dir/build.make src/CMakeFiles/protobuf-mutator.dir/build
make[2]: Entering directory '/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator'
[ 12%] Building CXX object src/CMakeFiles/protobuf-mutator.dir/mutator.cc.o
cd /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src && /usr/bin/c++  -I/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator -fno-exceptions -Werror -Wall -MD -MT src/CMakeFiles/protobuf-mutator.dir/mutator.cc.o -MF CMakeFiles/protobuf-mutator.dir/mutator.cc.o.d -o CMakeFiles/protobuf-mutator.dir/mutator.cc.o -c /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/mutator.cc
In file included from /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/mutator.cc:27:
/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/field_instance.h: In member function ‘std::string protobuf_mutator::ConstFieldInstance::DebugString() const’:
/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/field_instance.h:198:33: error: no matching function for call to ‘StrCat(const google::protobuf::FieldDescriptor&)’
  198 |     std::string s = absl::StrCat(*descriptor_);
      |                     ~~~~~~~~~~~~^~~~~~~~~~~~~~
In file included from /usr/include/absl/container/internal/layout.h:176,
                 from /usr/include/absl/strings/internal/cord_rep_ring.h:25,
                 from /usr/include/absl/strings/cord.h:89,
                 from /usr/include/google/protobuf/io/coded_stream.h:134,
                 from /usr/include/google/protobuf/any.pb.h:24,
                 from /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/port/protobuf.h:20,
                 from /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/mutator.h:28,
                 from /tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/mutator.cc:15:
/usr/include/absl/strings/str_cat.h:395:41: note: candidate: ‘template<class ... AV> std::string absl::lts_20230125::StrCat(const AlphaNum&, const AlphaNum&, const AlphaNum&, const AlphaNum&, const AlphaNum&, const AV& ...)’
  395 | ABSL_MUST_USE_RESULT inline std::string StrCat(
      |                                         ^~~~~~
/usr/include/absl/strings/str_cat.h:395:41: note:   template argument deduction/substitution failed:
/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator/src/field_instance.h:198:33: note:   candidate expects at least 5 arguments, 1 provided
  198 |     std::string s = absl::StrCat(*descriptor_);
      |                     ~~~~~~~~~~~~^~~~~~~~~~~~~~
/usr/include/absl/strings/str_cat.h:381:41: note: candidate: ‘std::string absl::lts_20230125::StrCat()’
  381 | ABSL_MUST_USE_RESULT inline std::string StrCat() { return std::string(); }
      |                                         ^~~~~~
/usr/include/absl/strings/str_cat.h:381:41: note:   candidate expects 0 arguments, 1 provided
/usr/include/absl/strings/str_cat.h:383:41: note: candidate: ‘std::string absl::lts_20230125::StrCat(const AlphaNum&)’
  383 | ABSL_MUST_USE_RESULT inline std::string StrCat(const AlphaNum& a) {
      |                                         ^~~~~~
/usr/include/absl/strings/str_cat.h:383:64: note:   no known conversion for argument 1 from ‘const google::protobuf::FieldDescriptor’ to ‘const absl::lts_20230125::AlphaNum&’
  383 | ABSL_MUST_USE_RESULT inline std::string StrCat(const AlphaNum& a) {
      |                                                ~~~~~~~~~~~~~~~~^
/usr/include/absl/strings/str_cat.h:387:34: note: candidate: ‘std::string absl::lts_20230125::StrCat(const AlphaNum&, const AlphaNum&)’
  387 | ABSL_MUST_USE_RESULT std::string StrCat(const AlphaNum& a, const AlphaNum& b);
      |                                  ^~~~~~
/usr/include/absl/strings/str_cat.h:387:34: note:   candidate expects 2 arguments, 1 provided
/usr/include/absl/strings/str_cat.h:388:34: note: candidate: ‘std::string absl::lts_20230125::StrCat(const AlphaNum&, const AlphaNum&, const AlphaNum&)’
  388 | ABSL_MUST_USE_RESULT std::string StrCat(const AlphaNum& a, const AlphaNum& b,
      |                                  ^~~~~~
/usr/include/absl/strings/str_cat.h:388:34: note:   candidate expects 3 arguments, 1 provided
/usr/include/absl/strings/str_cat.h:390:34: note: candidate: ‘std::string absl::lts_20230125::StrCat(const AlphaNum&, const AlphaNum&, const AlphaNum&, const AlphaNum&)’
  390 | ABSL_MUST_USE_RESULT std::string StrCat(const AlphaNum& a, const AlphaNum& b,
      |                                  ^~~~~~
/usr/include/absl/strings/str_cat.h:390:34: note:   candidate expects 4 arguments, 1 provided
make[2]: *** [src/CMakeFiles/protobuf-mutator.dir/build.make:90: src/CMakeFiles/protobuf-mutator.dir/mutator.cc.o] Error 1
make[2]: Leaving directory '/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator'
make[1]: *** [CMakeFiles/Makefile2:115: src/CMakeFiles/protobuf-mutator.dir/all] Error 2
make[1]: Leaving directory '/tmp/tmp.Ua56J9LBs4/libprotobuf-mutator'
make: *** [Makefile:136: all] Error 2
```

So it would be definitely be good if the libprotobuf-mutator build system would check for and require recent enough Abseil.

--- a/.github/workflows/cmake-multi-platform.yml
+++ b/.github/workflows/cmake-multi-platform.yml
@@ -36,7 +36,7 @@ jobs:
     steps:
     - uses: actions/checkout@v3
 
-    - name: Install Ninja
+    - name: Install Deps
       run: sudo apt install -y liblzma-dev ninja-build
 
     - name: Set reusable strings

--- a/cmake/external/googletest.cmake
+++ b/cmake/external/googletest.cmake
@@ -44,7 +44,7 @@ include (ExternalProject)
 ExternalProject_Add(${GTEST_TARGET}
     PREFIX ${GTEST_TARGET}
     GIT_REPOSITORY https://github.com/google/googletest.git
-    GIT_TAG v1.14.0
+    GIT_TAG v1.15.0
     UPDATE_COMMAND ""
     CMAKE_CACHE_ARGS -DCMAKE_C_COMPILER:FILEPATH=${CMAKE_C_COMPILER}
                      -DCMAKE_CXX_COMPILER:FILEPATH=${CMAKE_CXX_COMPILER}

--- a/cmake/external/protobuf.cmake
+++ b/cmake/external/protobuf.cmake
@@ -152,7 +152,7 @@ include (ExternalProject)
 ExternalProject_Add(${PROTOBUF_TARGET}
     PREFIX ${PROTOBUF_TARGET}
     GIT_REPOSITORY https://github.com/google/protobuf.git
-    GIT_TAG v27.1
+    GIT_TAG v29.3
     UPDATE_COMMAND ""
     CONFIGURE_COMMAND ${CMAKE_COMMAND} ${PROTOBUF_INSTALL_DIR}/src/${PROTOBUF_TARGET}
         -G${CMAKE_GENERATOR}

--- a/examples/libfuzzer/libfuzzer_bin_example.cc
+++ b/examples/libfuzzer/libfuzzer_bin_example.cc
@@ -55,7 +55,7 @@ DEFINE_BINARY_PROTO_FUZZER(const libfuzzer_example::Msg& message) {
       !std::isnan(message.optional_float()) &&
       std::fabs(message.optional_float()) > 1000 &&
       message.any().UnpackTo(&file) && !file.name().empty()) {
-    std::cerr << absl::StrCat(message) << "\n";
+    std::cerr << protobuf_mutator::protobuf::ShortFormat(message) << "\n";
     abort();
   }
 }

--- a/examples/libfuzzer/libfuzzer_example.cc
+++ b/examples/libfuzzer/libfuzzer_example.cc
@@ -55,7 +55,7 @@ DEFINE_PROTO_FUZZER(const libfuzzer_example::Msg& message) {
       !std::isnan(message.optional_float()) &&
       std::fabs(message.optional_float()) > 1000 &&
       message.any().UnpackTo(&file) && !file.name().empty()) {
-    std::cerr << absl::StrCat(message) << "\n";
+    std::cerr << protobuf_mutator::protobuf::ShortFormat(message) << "\n";
     abort();
   }
 }

--- a/src/field_instance.h
+++ b/src/field_instance.h
@@ -195,9 +195,9 @@ class ConstFieldInstance {
   const protobuf::FieldDescriptor* descriptor() const { return descriptor_; }
 
   std::string DebugString() const {
-    std::string s = absl::StrCat(*descriptor_);
-    if (is_repeated()) s += absl::StrCat("[", index_, "]");
-    return s + " of\n" + absl::StrCat(*message_);
+    std::string s = descriptor_->DebugString();
+    if (is_repeated()) s += "[" + std::to_string(index_) + "]";
+    return s + " of\n" + protobuf::ShortFormat(*message_);
   }
 
  protected:

--- a/src/mutator.cc
+++ b/src/mutator.cc
@@ -801,7 +801,7 @@ std::string Mutator::MutateUtf8String(const std::string& value,
 
 bool Mutator::IsInitialized(const Message& message) const {
   if (!keep_initialized_ || message.IsInitialized()) return true;
-  std::cerr << "Uninitialized: " << absl::StrCat(message) << "\n";
+  std::cerr << "Uninitialized: " << protobuf::ShortFormat(message) << "\n";
   return false;
 }
 

--- a/src/mutator_test.cc
+++ b/src/mutator_test.cc
@@ -410,8 +410,8 @@ bool Mutate(const protobuf::Message& from, const protobuf::Message& to,
   }
 
   ADD_FAILURE() << "Failed to get from:\n"
-                << absl::StrCat(from) << "\nto:\n"
-                << absl::StrCat(to);
+                << protobuf::ShortFormat(from) << "\nto:\n"
+                << protobuf::ShortFormat(to);
   return false;
 }
 

